retriever.py

```diff
@@ -13,20 +16,71 @@
 from preproc import get_functions
 from tokenizer import Tokenizer
 from transformers import AutoModelForCausalLM
+from gen_pr_items import PR_ITEMS
+
+REPO_ROOT = Path(__file__).resolve().parent
+
+
+class EmbeddingGenerator:
+    """
+    Generate embedding for an item based using the model
+    """
+
+    def __init__(self):
+        self.device = "cuda" if torch.cuda.is_available() else "cpu"
+        self.config = TDArgs()
+
+        generator = Llama.build(
+            ckpt_dir=os.path.expanduser(self.config.model_ckpt_dir),
+            tokenizer_path=os.path.expanduser(self.config.tokenizer_path),
+            max_seq_len=self.config.max_context_len,
+            max_batch_size=self.config.max_batch_size,
+            use_kv_cache=False,
+            model_parallel_size=1,
+        )
+        self.model = generator.model.to(self.device)
+        self.tokenizer = Tokenizer(self.config)
+
+    def gen_embedding(self, items: List[str]) -> List[Any]:
+        # Tokenizes and generates embedding for each item.  Returns a list of
+        # embeddings
+        self.model.eval()
+        with torch.autocast(
+            self.device
+        ):  # needed for cpu inference? something about half floats
+            with torch.no_grad():
+                embeddings = []
+                for item in items:
+                    tensor = torch.full(
+                        (1, self.config.max_context_len),
+                        self.tokenizer.pad_id,
+                        dtype=torch.long,
+                    )
+
+                    tokens = self.tokenizer.encode(item)
+                    tokens = tokens[: self.config.max_context_len]
+                    tensor[0, : len(tokens)] = torch.tensor(
+                        tokens, dtype=torch.long
+                    )
+
+                    tensor = tensor.to(self.device)
+
+                    _, func_embedding = self.model.forward(
+                        tensor, 0, output_last_hidden_state=True
+                    )
+                    pooled_embedding = torch.sum(func_embedding, dim=1)
+                    embeddings.append(pooled_embedding)
+                return embeddings
 
 
 class Retriever:
+    """Evaluate an embedding given an experiment"""
+
     def __init__(self, experiment_name):
+        self.device = "cuda" if torch.cuda.is_available() else "cpu"
```

Comment:
autocast not working with `cuda:0` for some reason
`RuntimeError: User specified autocast device_type must be 'cuda' or 'cpu'`